source/common/socket.h

```diff
@@ -0,0 +1,79 @@
+#include "envoy/registry/registry.h"
+#include "envoy/server/transport_socket_config.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/store.h"
+
+#include "api/client/socket.pb.h"
+#include "api/client/socket.pb.validate.h"
+
+namespace Nighthawk {
+
+using namespace Envoy; // We need this because of macro expectations.
+
+#define ALL_SOCKET_STATS(COUNTER)                                                                  \
+  COUNTER(closes)                                                                                  \
+  COUNTER(connects)                                                                                \
+  COUNTER(write_bytes)                                                                             \
+  COUNTER(writes)                                                                                  \
+  COUNTER(read_bytes)                                                                              \
+  COUNTER(reads)
+
+struct SocketStats {
+  ALL_SOCKET_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+class SocketConfigFactory
+    : public virtual Envoy::Server::Configuration::TransportSocketConfigFactory {
+public:
+  ~SocketConfigFactory() override = default;
+  std::string name() const override { return "NighthawkSocket"; }
+};
+
+class UpstreamSocketConfigFactory
```

Comment:
If I saw an interface called `TransportSocketConfigFactory` I would expect it to return something of type `TransportSocketConfig`. But this one returns `TransportSocketFactoryPtr`. 

Ideally it would be named `ProtoConfigTransportSocketFactoryFactory`.

Is there any way to use our class name to mitigate Envoy's naming, without also confusing people familiar with Envoy?